### PR TITLE
application: ServiceFactory and main Application class

### PR DESCRIPTION
**Pull request by lengau**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

CRAFT-1755

### CODE REVIEW

File: craft_application/app.py
Comment (lengau) [inline craft_application/app.py:1]: This was renamed to application.py.

File: craft_application/services/service_factory.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    def set_kwargs(
+        self,
+        service: str,
+        **kwargs: Any,  # noqa: ANN401 this is intentionally duck-typed.
+    ) -> None:
+        """Set up the keyword arguments to pass to a particular service class."""
+        self._service_kwargs[service] = kwargs
+
+    def __getattr__(self, service: str) -> services.BaseService:
```
Comment (tigarmo) [inline craft_application/services/service_factory.py:63]: Can you add a docstring either here (or possibly the class docs) linking this method and the PackageClass/LifecycleClass/etc attributes? I (think I) understand what it does but I think some documentation would be good
also I'm trying to decide if this is too magic for my taste but so far I think it's ok ;)

**Review by tigarmo — APPROVED** [review]
that's a lot of work! Very nice

File: craft_application/services/service_factory.py (hunk 1 above)
Comment (lengau) [reply craft_application/services/service_factory.py:62]: Done! I also realised that I can take advantage of the fact that `__getattr__` only runs when the attribute isn't found by other means, so I made some small changes to this method and the relevant test.

**Review by tigarmo — APPROVED** [review]
nice!

File: craft_application/application.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+)
+
+
+@final
+@dataclass(frozen=True)
+class AppMetadata:
+    """Metadata about a *craft application."""
+
+    name: str
+    summary: str | None = None
```
Comment (syu-w) [inline craft_application/application.py:49]: What's the target minimal python version? I think `A|B` is 3.10+?

File: craft_application/services/service_factory.py
Comment (syu-w) [inline craft_application/services/service_factory.py:1]: So this one is not a lazy import

**Review by syu-w — APPROVED** [review]
Good job!

File: craft_application/application.py (hunk 2 above)
Comment (lengau) [reply craft_application/application.py:49]: [`from __future__ import annotations`](https://docs.python.org/3.8/library/__future__.html#__future__.annotations) brings it into Python 3.7+ Combine with ruff's `UP007` rule and it'll automatically convert once you use postponed annotations (which I originally included for return values).

File: craft_application/services/service_factory.py
Comment (lengau) [reply craft_application/services/service_factory.py:1]: I'm not clear what you mean here?

Comment (syu-w) [reply craft_application/services/service_factory.py:1]: I mean this one seems lazy create the service object, not lazy doing
```
from XXX import XXService
service = XXService()
...
return service
```

Comment (lengau) [reply craft_application/services/service_factory.py:1]: The lazy-loading part is the service instantiation, not the module import

**Comment by sergiusens** [comment]
why isn't this getting merged?

**Pull request merged**